src/hotspot/share/ci/ciObject.cpp

```diff
@@ -168,6 +168,26 @@ jobject ciObject::constant_encoding() {
   return handle();
 }
 
+// ------------------------------------------------------------------
+// ciObject::check_constant_value_cache()
+//
+// Cache constant value lookups to ensure that consistent values are observed
+// during compilation because fields may be (re-)initialized concurrently.
+ciConstant ciObject::check_constant_value_cache(int off, ciConstant val) {
+  if (_constant_values == nullptr) {
+    Arena* arena = CURRENT_ENV->arena();
+    _constant_values = new (arena) GrowableArray<ConstantValue>(arena, 0, 0, ConstantValue());
+  }
+  for (int i = 0; i < _constant_values->length(); ++i) {
+    ConstantValue cached_val = _constant_values->at(i);
+    if (cached_val.off() == off) {
```

Comment:
I'd also check that basic types matches (either with a runtime check or an assert).

Reply:
I added an assert that immediately triggered because `ciArray::element_value_impl` creates a `ciConstant` with basic type `T_OBJECT` for a `T_ARRAY` array element. I fixed that code.